Rollback transaction if incorrect isolation level is passed

zoso10 opened the pull request:
### Summary
This fixes an issue where a transaction is left open if an incorrect
isolation level is passed as an option. eg:
```ruby
Table.transaction(isolation: "foobar") do
  # ...
end
```
This behaviour can have the side effect of unknowingly leaving lock
entries in the database until another transaction is committed or rolled
back.

This solution rescues the `KeyError`, executes a transaction rollback, then
re-raises the `KeyError`, leaving behaviour the way it currently is,
except rolling back the failed transaction.


### Other Information
This is an issue only for postgresql databases since setting transaction
isolation levels *must* happen inside transactions. In mysql, however,
specifying isolation levels outside transactions will apply it to the
next transaction that is opened; so we don't need to change the mysql
adapter.

zoso10 committed:
Rollback transaction if incorrect isolation level is passed

This fixes an issue where a transaction is left open if an incorrect
isolation level is passed as an option. eg:
`Table.transaction(isolation: "foobar") { #... }`
This behaviour can have the side effect of unknowingly leaving lock
entries in the database until another transaction is committed or rolled
back.

This is an issue only for postgresql databases since setting transaction
isolation levels must happen inside transactions. In mysql, however,
specifying isolation levels outside transactions will apply it to the
next transaction that is opened; so we don't need to change the mysql
adapter.

This solution rescues the `KeyError`, executes a transaction rollback, then
re-raises the `KeyError`, leaving behaviour the way it currently is,
except rolling back the failed transaction.

matthewd commented:
I think we should do the rollback for any exception raised by `execute` (though not the begin).

We also need a test to match.

zoso10 commented:
the method `#withing_new_transaction` in the abstract adapter catches all `Exception`s and has some logic to rollback transactions. the issue there is that creating the `RealTransaction` object fails in the initializer so we never get an object causing the truthy conditional to evaluate false. i couldn't deduce what the original intention was for checking object presence so i shied away from editing the code there.

zoso10 commented:
@matthewd I changed the original approach and added a condition in the `rescue` of `ConnectionAdapter::TransactionManager#within_new_transaction`. if we add it specifically as a second branch we avoid rescuing `KeyError`s inside the transaction since the `transaction` object will exist and trigger the first branch as is current behaviour. This guarantees we only `rescue` when the `KeyError` is raised during opening the transaction, specifically looking up the isolation level. It involves a typecheck on the error object, which imo is not ideal, but functional nonetheless.
I created a test case for this as well. I created it as a `PostgreSQLTestCase` since the sql used to check for idle transactions is postgres specific.

oniofchaos commented:
👍

simi commented:
Wouldn't be easier to check on `transaction` call if level is valid and raise before transaction is actually open?
